CONTRIBUTING.md

```diff
@@ -23,3 +23,34 @@ use GitHub pull requests for this purpose. Consult [GitHub Help] for more
 information on using pull requests.
 
 [GitHub Help]: https://help.github.com/articles/about-pull-requests/
+
+## Building Dagger
+
+Dagger is built with [`bazel`](https://bazel.build).
+
+### Building Dagger from the command line
+
+* Go to the [Bazel installation page](https://docs.bazel.build/versions/master/install.html) and install Bazel.
+  * Learn more about Bazel targets [here](https://docs.bazel.build/versions/master/build-ref.html).
+* Clone the Dagger repository: `git clone git@github.com:google/dagger.git dagger`
+* Build the Dagger project with `bazel build`
+  * If you get the following error: `ERROR: missing input file '@androidsdk//:build-tools/26.0.2/aapt'`, install the missing build tools version with sdkmanager.
+* You can install the Dagger libraries in your **local maven repository** by running the `./util/install-local-snapshot.sh` script.
+  * It will build the libraries and install them with a `LOCAL-SNAPSHOT` version.
+* You can **run all tests** with `bazel test //...` and a single test with the path and class name, e.g. `bazel test ///javatests/dagger/internal/codegen:KeyFactoryTest`.
```

Comment:
This now comes up with the following error on fresh clone:

```console
❯ bazel test //...
INFO: Invocation ID: dbd34874-4f19-42d2-b832-9bf54c82a3b1
ERROR: /private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/BUILD:597:1: Traceback (most recent call last):
	File "/private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/BUILD", line 597
		internal_gen_well_known_protos_java(srcs = WELL_KNOWN_PROTOS)
	File "/private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/protobuf.bzl", line 266, in internal_gen_well_known_protos_java
		Label(("%s//protobuf_java" % REPOSITOR...))
	File "/private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/protobuf.bzl", line 266, in Label
		REPOSITORY_NAME
The value 'REPOSITORY_NAME' has been removed in favor of 'repository_name()', please use the latter (https://docs.bazel.build/versions/master/skylark/lib/native.html#repository_name). You can temporarily allow the old name by using --incompatible_package_name_is_a_function=false
DEBUG: /private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/bazel_tools/tools/cpp/osx_cc_configure.bzl:135:13: wrapped_clang failed to generate. This shouldn't cause problems, but please file an issue at https://github.com/bazelbuild/bazel/issues with the following:
return code 256, stderr: Process terminated by signal 15, stdout:
ERROR: /private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/BUILD:380:1: Target '@com_google_protobuf//:android' contains an error and its package is in error and referenced by '@com_google_protobuf//:protoc'
ERROR: /private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/BUILD:380:1: Target '@com_google_protobuf//:windows' contains an error and its package is in error and referenced by '@com_google_protobuf//:protoc'
ERROR: /private/var/tmp/_bazel_zsweers/1825c3febee87f83f0190e5ba08c9307/external/com_google_protobuf/BUILD:380:1: Target '@com_google_protobuf//:windows_msvc' contains an error and its package is in error and referenced by '@com_google_protobuf//:protoc'
ERROR: /Users/zsweers/dev/java/contributing/dagger/javatests/dagger/grpc/functional/server/BUILD:16:1: every rule of type proto_library implicitly depends upon the target '@com_google_protobuf//:protoc', but this target could not be found because of: Target '@com_google_protobuf//:protoc' contains an error and its package is in error
ERROR: Analysis of target '//javatests/dagger/grpc/functional/server:protos' failed; build aborted: Analysis failed
INFO: Elapsed time: 6.382s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (9 packages loaded, 489 targets configured)
FAILED: Build did NOT complete successfully (9 packages loaded, 489 targets configured)
    Fetching @remotejdk_macos; fetching 4s
```

This is because dagger uses a legacy API. To work around this, use `--incompatible_package_name_is_a_function=false`

```console
❯ bazel test //... --incompatible_package_name_is_a_function=false
```

Reply:
I think this could be covered by adding this to `.bzlrc`? CC @ronshapiro 

```
build --incompatible_package_name_is_a_function=false
```

Reply:
After running that, I get a new issue that seems like a bug on current master?

```console
❯ bazel test //... --incompatible_package_name_is_a_function=false
INFO: Invocation ID: 4c52fd9e-a8f2-45c6-8c17-4a7a5a4fff6b
INFO: Analysed 497 targets (1 packages loaded, 138 targets configured).
INFO: Found 111 targets and 386 test targets...
INFO: From Compiling external/com_google_protobuf/src/google/protobuf/compiler/js/embed.cc [for host]:
external/com_google_protobuf/src/google/protobuf/compiler/js/embed.cc:37:12: warning: unused variable 'output_file' [-Wunused-const-variable]
const char output_file[] = "well_known_types_embed.cc";
           ^
1 warning generated.
ERROR: /Users/zsweers/dev/java/contributing/dagger/javatests/dagger/android/support/functional/BUILD:20:1: Building javatests/dagger/android/support/functional/libfunctional.jar (13 source files) and running annotation processors (AndroidProcessor, ComponentProcessor) failed (Exit 1)
bazel-out/darwin-fastbuild/bin/javatests/dagger/android/support/functional/_javac/functional/libfunctional_sourcegenfiles/dagger/android/support/functional/UsesGeneratedModulesApplication_ApplicationModule_ContributeTestActivityWithScopeInjector.java:16: error: [strict] Using type javax.annotation.Generated from an indirect dependency (TOOL_INFO: "@javax_annotation_jsr250_api//:javax_annotation_jsr250_api"). See command below **
@Generated("dagger.android.processor.AndroidProcessor")
 ^
 ** Please add the following dependencies:
  @javax_annotation_jsr250_api//:javax_annotation_jsr250_api to //javatests/dagger/android/support/functional
 ** You can use the following buildozer command:
buildozer 'add deps @javax_annotation_jsr250_api//:javax_annotation_jsr250_api' //javatests/dagger/android/support/functional

INFO: Elapsed time: 157.949s, Critical Path: 108.35s
INFO: 482 processes: 346 darwin-sandbox, 136 worker.
FAILED: Build did NOT complete successfully
```

So running that buildozer command, turns out that's not part of bazel's homebrew installation

```console
❯ buildozer 'add deps @javax_annotation_jsr250_api//:javax_annotation_jsr250_api' //javatests/dagger/android/support/functional
zsh: command not found: buildozer
```

Nor does it exist in the bazel tap or on homebrew in general

```console
❯ brew search buildozer
No formula or cask found for "buildozer".
```

Googling it turns up https://github.com/bazelbuild/buildtools/tree/master/buildozer. It's a Go thing

So I install it

```console
❯ go get github.com/bazelbuild/buildtools/buildozer
```

And re-run

```console
❯ buildozer 'add deps @javax_annotation_jsr250_api//:javax_annotation_jsr250_api' //javatests/dagger/android/support/functional
fixed /Users/zsweers/dev/java/contributing/dagger/javatests/dagger/android/support/functional/BUILD
```

Yielding this patch

```patch
diff --git a/javatests/dagger/android/support/functional/BUILD b/javatests/dagger/android/support/functional/BUILD
index ffb93618..a67e7027 100644
--- a/javatests/dagger/android/support/functional/BUILD
+++ b/javatests/dagger/android/support/functional/BUILD
@@ -33,6 +33,7 @@ android_library(
         "@androidsdk//com.android.support:appcompat-v7-25.0.0",
         "@androidsdk//com.android.support:support-fragment-25.0.0",
         "@google_bazel_common//third_party/java/guava",
+        "@javax_annotation_jsr250_api",
     ],
 )
 

```

After this I was able to successfully build all tests

Reply:
That flag is removed in recent Bazel version https://github.com/bazelbuild/bazel/issues/7295.

Reply:
> I think this could be covered by adding this to `.bzlrc`? CC @ronshapiro
> 
> ```
> build --incompatible_package_name_is_a_function=false
> ```

ERROR: Unrecognized option: --incompatible_package_name_is_a_function=false
Build label: 0.24.1

Reply:
My recent changes to bring us up to 0.24.1 should make these go away